hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java

```diff
@@ -0,0 +1,707 @@
+
+	private final boolean ansiSequence;
+	private final boolean cascadeConstraints;
+	private final boolean useLocalTime;
+
+	private final SequenceInformationExtractor sequenceInformationExtractor;
+	private final String querySequenceString;
+
+	public H2LegacyDialect(DialectResolutionInfo info) {
+		this( parseVersion( info ) );
+		registerKeywords( info );
+	}
+
+	public H2LegacyDialect() {
+		this( SimpleDatabaseVersion.ZERO_VERSION );
+	}
+
+	public H2LegacyDialect(DatabaseVersion version) {
+		super(version);
+
+		// https://github.com/h2database/h2database/commit/b2cdf84e0b84eb8a482fa7dccdccc1ab95241440
+		limitHandler = version.isSameOrAfter( 1, 4, 195 )
+				? OffsetFetchLimitHandler.INSTANCE
+				: LimitOffsetLimitHandler.INSTANCE;
+
+		if ( version.isBefore( 1, 2, 139 ) ) {
+			LOG.unsupportedMultiTableBulkHqlJpaql( version.getMajor(), version.getMinor(), version.getMicro() );
+		}
+
+//		supportsTuplesInSubqueries = version.isSameOrAfter( 1, 4, 198 );
+
+		// Prior to 1.4.200 there was no support for 'current value for sequence_name'
+		// After 2.0.202 there is no support for 'sequence_name.nextval' and 'sequence_name.currval'
+		ansiSequence = version.isSameOrAfter( 1, 4, 200 );
+
+		// Prior to 1.4.200 the 'cascade' in 'drop table' was implicit
+		cascadeConstraints = version.isSameOrAfter( 1, 4, 200 );
+		// 1.4.200 introduced changes in current_time and current_timestamp
+		useLocalTime = version.isSameOrAfter( 1, 4, 200 );
+
+		if ( version.isSameOrAfter( 1, 4, 32 ) ) {
+			this.sequenceInformationExtractor = version.isSameOrAfter( 1, 4, 201 )
+					? SequenceInformationExtractorLegacyImpl.INSTANCE
+					: SequenceInformationExtractorH2DatabaseImpl.INSTANCE;
+			this.querySequenceString = "select * from INFORMATION_SCHEMA.SEQUENCES";
+		}
+		else {
+			this.sequenceInformationExtractor = SequenceInformationExtractorNoOpImpl.INSTANCE;
+			this.querySequenceString = null;
+		}
+	}
+
+	private static DatabaseVersion parseVersion(DialectResolutionInfo info) {
+		return DatabaseVersion.make( info.getMajor(), info.getMinor(), parseBuildId( info ) );
+	}
+
+	private static int parseBuildId(DialectResolutionInfo info) {
+		final String databaseVersion = info.getDatabaseVersion();
+		if ( databaseVersion == null ) {
+			return 0;
+		}
+
+		final String[] bits = databaseVersion.split("[. ]");
+		return bits.length > 2 ? Integer.parseInt( bits[2] ) : 0;
+	}
+
+	@Override
+	public boolean getDefaultNonContextualLobCreation() {
+		// http://code.google.com/p/h2database/issues/detail?id=235
+		return true;
+	}
+
+	@Override
+	public boolean supportsStandardArrays() {
+		return getVersion().isSameOrAfter( 2 );
+	}
+
+	@Override
+	protected String columnType(int sqlTypeCode) {
+		switch ( sqlTypeCode ) {
+			// prior to version 2.0, H2 reported NUMERIC columns as DECIMAL,
+			// which caused problems for schema update tool
+			case NUMERIC:
+				return getVersion().isBefore( 2 ) ? columnType( DECIMAL ) : super.columnType( sqlTypeCode );
```

Comment:
H2 1.4.197, 1.4.198, 1.4.199, and 1.4.200 return either `Types.DECIMAL` or `Types.NUMERIC` depending on configuration, `Types.DECIMAL` is used by default. But I think presence of this configuration option can be ignored, it is rarely used.

H2 2.0.202 and newer versions return `Types.DECIMAL` for `DECIMAL` / `DEC` and `Types.NUMERIC` for `NUMERIC` columns. But their real type is always `NUMERIC`, there is no specialized implementation for `DECIMAL` / `DEC`.

I don't understand why this condition is required for legacy dialect, it looks like it isn't expected to be used with H2 2.\*.\*.

Reply:
If you have ideas to improve the dialect, please create an issue for that and provide a PR